parsl/executors/status_handling.py

```diff
@@ -0,0 +1,155 @@
+import logging
+import threading
+from abc import abstractmethod
+from concurrent.futures import Future
+from typing import List, Any, Dict
+
+import parsl  # noqa F401
+from parsl.executors.base import ParslExecutor
+from parsl.providers.provider_base import JobStatus, ExecutionProvider, JobState
+
+logger = logging.getLogger(__name__)
+
+
+class StatusHandlingExecutor(ParslExecutor):
+    def __init__(self, provider):
+        super().__init__()
+        self._provider = provider  # type: ExecutionProvider
+        # errors can happen during the sumbit call to the provider; this is used
+        # to keep track of such errors so that they can be handled in one place
+        # together with errors reported by status()
+        self._simulated_status = {}
+        self._executor_bad_state = threading.Event()
+        self._executor_exception = None
+        self._generated_job_id_counter = 1
+        self._tasks = {}  # type: Dict[str, Future]
+
+    def _make_status_dict(self, job_ids: List[Any], status_list: List[JobStatus]) -> Dict[Any, JobStatus]:
+        """Given a list of job ids and a list of corresponding status strings,
+        returns a dictionary mapping each job id to the corresponding status
+
+        :param job_ids: the list of job ids
+        :param status_list: the list of job status strings
+        :return: the resulting dictionary
+        """
+        if len(job_ids) != len(status_list):
+            raise IndexError("job id list and status string list differ in size")
+        d = {}
+        for i in range(len(job_ids)):
+            d[job_ids[i]] = status_list[i]
+
+        return d
+
+    def _set_provider(self, provider: ExecutionProvider):
+        self._provider = provider
+
+    @property
+    def status_polling_interval(self):
+        if self._provider is None:
+            return 0
+        else:
+            return self._provider.status_polling_interval
+
+    @abstractmethod
+    def _get_job_ids(self) -> List[Any]:
+        raise NotImplementedError("Classes inheriting from StatusHandlingMixin must implement"
+                                  "_get_job_ids()")
+
+    def _fail_job_async(self, job_id: Any, message: str):
+        """Marks a job that has failed to start but would not otherwise be included in status()
+        as failed and report it in status()
+        """
+        if job_id is None:
+            job_id = "block-{}".format(self._generated_job_id_counter)
```

Comment:
put "failed-block" or something in this string so that people wondering why they get two different block ID formats later get more of a clue.